### DISCUSSION TITLE
New flat config poses challenges for DefinitelyTyped

### DISCUSSION THREAD

**Discussion by jakebailey** [eslint/eslint#17853] [General] [answered]
Sending this over per https://twitter.com/andhaveaniceday/status/1734628817787056253

DefinitelyTyped (https://github.com/DefinitelyTyped/DefinitelyTyped) is a monorepo with over 9000 packages. We use ESLint heavily, using typescript-eslint plus our own custom rules.

The new flat config system poses new challenges for us. Here are the biggest issues we'll face:

## The config must be named `eslint.config.js`

DefinitelyTyped packages describe real npm packages; their `package.json` `type` field must match the package they describe in order to correctly model module resolution. Becuase the config files must be named `eslint.config.js`, the format of this file depends on an external factor that we cannot control.

The DefinitelyTyped repo root itself is `type=module` (for our scripts). This means that our root ESLint config would be in MJS format. But, 9124 our of the 9179 packages are `type=commonjs`. This means that for nearly every package in the repo that needs to override some rules, they are unable to extend the root config without writing:

```js
module.exports = (async () => {
    const root = await import("../../eslint.config.js");
    return [
        ...root,
        { ... }
    ];
})();
```

This is very heavyweight for disabling one or two rules. Right now, the JSON hierarchy means that a rule being diabled is as simple as:

```json
{
    "rules": {
        "some-rule": "off"
    }
}
```

If `mjs` or `cjs` were supported, we'd be able to pick one format for the entire repo (mjs!) and not have to write extra confusing code. (In my own executable-config-using project hereby, I support both `.mjs` and `.js`, but error when more than one are present; I think that is a good way to avoid ambiguity.)

I am operating under the assumption that child eslint configs need to import and spread their parent's configs (and it's not like the old JSON rules where it merged with files above). The new docs don't mention "extending" at all (https://eslint.org/docs/latest/use/configure/configuration-files-new), so this was unclear to me. Perhaps this isn't needed and configs still cascade down?

## `eslint.config.js` is exectuable

`eslint.config.js` is executable. Right now, we statically analyze `tslint.json` (and intended to do the same for `.eslintrc.json`), both because we want to prevent users from doing the wrong thing in their DT contributions, but also so that we can perform rule migrations across hundreds of config files.

The new config format is much less conducive to static analysis. Before, we could just load JSON. Now, we'd need to parse JS, ensure it's in an expected format, etc.

Additionally, the fact that these configurations are executable is a security problem for us. Most DefinitelyTyped PRs are automatically merged. Not all config changes require manual review, but now, there's theoretically a way that someone could commit malicious code into the exectuable JS format and make it into our CI. We're going to have to ensure that we never run ESLint in any trusted environment... but even that may not be good enough, because one user could merge a config that runs code on any other DT contibutor's machine.

It's possible that we could write some required CI check that verifies that the contents of the file are in a specific format, but it is a little frustrating to have no non-code-executing method to do this. (It's made extra hard due to the `.js` only extension; now we'd need two formats.)

I don't think there's any sort of workaround possible for this, if you're committed to JS-only configuration, but I wanted to note that it's problematic anyway.

---

That being said, it's not all negative with flat config; being able to get rid of project-local resolution and replacing our plugin preset with one that directly imports the plugins we use is a simplification I am looking forward to.

I do have other comments regarding `.js` not _really_ driving adoption of ESM, but those may only be in response to other Twitter users as I haven't seen anything here that suggests that the point of `.js` is to get more people to use ESM.

**Comment by bradzacher** [comment]
One suggestion is that you can require that users define a `.eslintrc.json` and then in your eslint tooling you can load that automatically via `FlatCompat` eg

```ts
const { FlatCompat } = require('@eslint/eslintrc'); 

const compat = new FlatCompat({ resolvePluginsRelativeTo: __dirname });

module.exports = [compat.config(require('./.eslintrc.json'))];
```

Or alternately you could just have them author the flat config format as json and then similarly require it in the flat config:
```ts
module.exports = [
  ...require('../eslint.config.js'),
  ...require('./eslint.config.json'),
];
```

That would at least allow you to have a locked down, config-only json file config without the execution vulnerability.
You could have a CI step which enforces each `eslint.config.js` exactly matches the required code so users cannot add other runtime code to the config files.

At least let's you work around the security side of your issues.

**Reply by jakebailey** [reply]
This is the mechanism I was thinking we could use, though forcing a specific format in JS is not... too hard, but what you've written is certainly easier.

**Reply by jakebailey** [reply]
Note that this doesn't have a benefit for security per se; we still have to have the JS code there so therefore it must have a specific format, otherwise people can still sneak stuff into it... And so that depends on us never accidentally merging something bad if CI isn't working or we just plain miss it.

**Reply by jakebailey** [reply]
Now... We could make the repo root the only allowed config, and then make a giant loop there that reads every other JSON config and adds it as an override. But that sounds quite slow.

**Reply by bradzacher** [reply]
Could have a CI step that essentially enforces that each `.js` config matches the expected content before executing it - thus enforcing the user isn't playing funny buggers.

**Reply by jakebailey** [reply]
Yes, for sure, but that unfortunately doesn't help contributors who are cloning DT and running our tooling from getting pwn'd.

**Reply by jakebailey** [reply]
But, I can understand that it's our problem to not merge malicious code... We'd need to invent a class of "required file that must match a specific state and can be merged so long as it matches", and then be really, really careful.

It just is frustrating that we have to deal with this vector at all. Of course, I see no world in which JSON configs are brought back, given the direction of the project, so I've made my peace with that.

**Comment by fasttime** [comment]
> This means that for nearly every package in the repo that needs to override some rules, they are unable to extend the root config without writing:
> 
> ```js
> module.exports = async () => {
>     const root = await import("../../eslint.config.js");
>     return [
>         ...root,
>         { ... }
>     ];
> };
> ```

Just a note that exporting an async function in the config file doesn't work. As explained in the [documentation](https://eslint.org/docs/latest/use/configure/configuration-files-new#configuration-file), in CommonJS the config should export a promise. For example:

```js
module.exports = (async () => {
    const root = await import("../../eslint.config.js");
    return [
        ...root,
        { ... }
    ];
})();
```

**Reply by jakebailey** [reply]
Ah, yes, sorry, I meant to make it a Promise but misremembered.

**Comment by nzakas** [comment]
> ## The config must be named `eslint.config.js`
> 
> DefinitelyTyped packages describe real npm packages; their `package.json` `type` field must match the package they describe in order to correctly model module resolution. Becuase the config files must be named `eslint.config.js`, the format of this file depends on an external factor that we cannot control.

Your config file doesn't have to be called `eslint.config.js` -- this is just the default one that ESLint searches for. More to come down below...

> The DefinitelyTyped repo root itself is `type=module` (for our scripts). This means that our root ESLint config would be in MJS format. But, 9124 our of the 9179 packages are `type=commonjs`.

Yeah, unfortunately this is going to be heavier than it was previously, but also, it will be a lot less error-prone. The cascading nature of eslintrc meant that complexity exploded exponentially with each additional config file in the cascade. The way we typically suggest people address a use case like yours (and I'm not sure this will work for you in a monorepo, but has been our advice) is to have a single config file at the root of your project and then use `files` to target specific configurations at specific directories. So instead of a core config file and then directory-specific override files, just have one root config file that specifies the directory-level overrides.

> If `mjs` or `cjs` were supported, we'd be able to pick one format for the entire repo (mjs!) and not have to write extra confusing code. (In my own executable-config-using project hereby, I support both `.mjs` and `.js`, but error when more than one are present; I think that is a good way to avoid ambiguity.)

You can use the `-c` CLI flag to pass in any filename you want, including `eslint.config.cjs` and `eslint.config.mjs`. 😄 


> I am operating under the assumption that child eslint configs need to import and spread their parent's configs (and it's not like the old JSON rules where it merged with files above). The new docs don't mention "extending" at all (https://eslint.org/docs/latest/use/configure/configuration-files-new), so this was unclear to me. Perhaps this isn't needed and configs still cascade down?

Yes. The idea is that you place everything into an array where the [cascade is obvious](https://eslint.org/docs/latest/use/configure/configuration-files-new#cascading-configuration-objects). We're basically taking all of the magic behind-the-scenes stuff we did in eslintrc and make it explicit in flat config.
 
> ## `eslint.config.js` is exectuable
> 
> `eslint.config.js` is executable. Right now, we statically analyze `tslint.json` (and intended to do the same for `.eslintrc.json`), both because we want to prevent users from doing the wrong thing in their DT contributions, but also so that we can perform rule migrations across hundreds of config files.
> I don't think there's any sort of workaround possible for this, if you're committed to JS-only configuration, but I wanted to note that it's problematic anyway.

Yeah, this is the downside of using JavaScript as a config file format, and it would have been useful to get this type of feedback two years ago when we were doing the RFC, but alas, here we are and that shipped has sailed.

As others mentioned, you do have the option of using JSON files to store any serializable configuration information in a non-executable way and then load it into your config file. Once again, this is one of the benefits of flat config: you can load data into your config file in any way that you want without asking permission from us. Y

> I do have other comments regarding `.js` not _really_ driving adoption of ESM, but those may only be in response to other Twitter users as I haven't seen anything here that suggests that the point of `.js` is to get more people to use ESM.

Yeah, I'm not sure who was saying that, but it didn't come from me or anyone else on the team. Standardizing on `.js` has nothing to do with "driving adoption of ESM" or whatever else people are saying on the internet. The point is that the overwhelming majority of those who responded to the ESLint user survey are using ESM primarily. CommonJS is dying out and so I don't see the need to bend over backwards to continue supporting it. With eslintrc, we supported multiple different filenames and formats, and that's a decision I regret. We should have just stuck with one. I didn't want to make the same mistake with flat config.

And as I already mentioned, you can use `eslint.config.cjs` or `eslint.config.mjs` with the `-c` command line option if you really feel strongly that this will make your ESLint experience better. 

I hope this helps and I appreciate the detailed feedback.

**Reply by jakebailey** [reply]
> Your config file doesn't have to be called `eslint.config.js` -- this is just the default one that ESLint searches for.

I considered it, but passing `-c` doesn't seem like a great option. ESLint works in the editor on DT and attempting to get everyone a correct config that works for them by default is not easy. The typical workflow on DT is for a user to only open one package in their editor. VS Code will not go look above, which means that if we want to have a different config name, all 9000 packages will need copies of the same `.vscode/settings.json`. We typically _don't_ commit those, because people have different preferences.

I would much rather use `eslint.config.js` and have it be different per package and very large in most packages than to solve the editor problem.

---

> The way we typically suggest people address a use case like yours (and I'm not sure this will work for you in a monorepo, but has been our advice) is to have a single config file at the root of your project and then use files to target specific configurations at specific directories. So instead of a core config file and then directory-specific override files, just have one root config file that specifies the directory-level overrides.

This doesn't feel super feasible for DT. We have 9000 packages. 2777 of them have their own eslint configs. This combined with the fact that the automation doesn't allow people to self-merge code that modifies files outside of their package means that having one big shared config is not super great.

I can see a world in which I do the hack I mentioned in the other thread and write one config that then loops over all 9000 dirs and then merges their configs together. That's roughly equivalent to using one config, except very likely slow, but maybe it's viable if I can show that the perf is okay on all OSs.

---

> Yeah, this is the downside of using JavaScript as a config file format, and it would have been useful to get this type of feedback two years ago when we were doing the RFC, but alas, here we are and that shipped has sailed.

2 years ago I was barely on the TypeScript team, so I apologize that I could not give that feedback. I'd love to create some channel by which this kind of feedback could be given without people needing to remember to look, though.

I know that TS itself has very manifestly _not_ doing executable configs for the same reasons, so I suspect that we would have advised that way 2 years ago too. Not sure if anyone else noted that in prior discussion.

---

> As others mentioned, you do have the option of using JSON files to store any serializable configuration information in a non-executable way and then load it into your config file. Once again, this is one of the benefits of flat config: you can load data into your config file in any way that you want without asking permission from us. 

The trouble is that there's no solid way to ensure that this never happens, but, perhaps the "one config that attempts to load all of the other configs" method is enough to make it very obvious when someone tries to commit something bad, along with a rule that staight up bans `eslint.config.js` anywhere in the repo.

---

> With eslintrc, we supported multiple different filenames and formats, and that's a decision I regret. We should have just stuck with one. I didn't want to make the same mistake with flat config.

Even still, I would strongly suggest allowing `.cjs` and `.mjs` too, and then just erroring if more than one are present in a directory. That to me has very little mental overhead and required documentation, along with a pretty simple implementation.

**Reply by Stanzilla** [reply]
Agree with Jake here, if the CLI accepts the files, they should work with auto-detect as well. Otherwise it's just more friction for little gain.

**Reply by karlhorky** [reply]
> you can use eslint.config.cjs or eslint.config.mjs with the -c command line option

Is this configurable out of the box with all ESLint tooling eg. the VS Code ESLint extension? A command line option seems pretty low discoverability for users compared to out of the box support for at least `.mjs` (gut feel says most users who are already in the challenging situation of trying to transition to ESM will not find it and/or will not be able to use it)

> CommonJS is dying out

Only 18.25% of ecosystem has moved to ESM or dual as per recent graph by @wooorm 

- https://twitter.com/wooorm/status/1727343107740704801

This is going to be a long migration and a super slow process, and any 1st-party tools support (not a config option, which requires extra work to configure across whole toolchain) to help the community migrate is very helpful!

**Reply by bradzacher** [reply]
> You can use the -c CLI flag to pass in any filename you want, including eslint.config.cjs and eslint.config.mjs.

Sure on the CLI that's fine and easy to control for all repo consumers via package.json scripts.

But each IDE is configured differently (if at all) and not all of them support a checked-in config file to configure the ESLint functionality.

Two quick examples:
- vscode
  - has the `eslint.experimental.useFlatConfig` flag which must be set in order to enable flat config support.
  - has the `eslint.options` setting which can be used to pass options to the `ESLint` constructor which should let you pass a custom config filename.
  - supports a checked-in config file
- Intellij
  - will automatically use flat config if it detects a `eslint.config.js`.
  - has no support for custom flat config file names right now.

At Canva we have over a thousand engineers that use both IntelliJ and VSCode.
For legacy reasons we can't use a `.vscode/settings.json` to add repo config - so any config change requires a heavy and disruptive change management process.

Eg if we used `.mjs` we'd put our vscode users though massive, disruptive pain and our IJ users would no longer be able to lint in their IDE. Both of these make it a non-starter for us - so if we want ESM we'll need to work around it.

For this same reason we currently can't experiment with a flat config - because we can't get user's IDEs to recognise it without the disruptive config change management processes.

In comparison when we switched from `.eslintrc.json` to `.eslintrc.js` it was completely invisible to all users and thus completely painless.

I'd love to migrate us now - because importing plugins would greatly simplify our setup with our internal-only plugins. But keeping the IDE working is infinitely higher priority than simplifying the setup.

---

Also worth noting that currently our repo is set to CJS because (a) we were on node 14 until about 6m ago and (b) there's >60k files that need to be validated in "ESM" mode. We'd like to migrate fully to ESM but it's a big job with "low value" so hard to prioritise against other work.
But using ESM for all new files is easy to do and saves on future migration size. Eg all the infra we have built since upgrading from Node 14 has been in `.mjs` files. Which is why we'd love to use `.mjs` for this config too.

**Comment by nzakas** [comment]
@jakebailey

> I considered it, but passing -c doesn't seem like a great option. ESLint works in the editor on DT and attempting to get everyone a correct config that works for them by default is not easy. The typical workflow on DT is for a user to only open one package in their editor. VS Code will not go look above, which means that if we want to have a different config name, all 9000 packages will need copies of the same .vscode/settings.json. We typically don't commit those, because people have different preferences.

Ah, this is *super* interesting. I didn't realize people worked on one package at a time in the editor rather than working with the entire repo. I'll admit my knowledge of managing monorepos is fairly elementary as I've never had to use one. So it sounds like you're saying that each package is effectively maintained separately without any overriding management?

> I know that TS itself has very manifestly not doing executable configs for the same reasons, so I suspect that we would have advised that way 2 years ago too. Not sure if anyone else noted that in prior discussion.

I just checked and the RFC was initially posted in 2019 :sweat::
https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification

Pull request:
https://github.com/eslint/rfcs/pull/9

In any event, the tradeoffs between static vs. executable were discussed (though probably without a security focus), but I think the RFC lays out pretty clearly why static config files were just not working for us. 

> Even still, I would strongly suggest allowing .cjs and .mjs too, and then just erroring if more than one are present in a directory. That to me has very little mental overhead and required documentation, along with a pretty simple implementation.

The problem with this approach is that we then need to search for all three files in every directory that we search (three `fs.stat()` calls). That's why for eslintrc we just took the first one that was found. 

That said, I think some good points we hadn't considered were mentioned in this thread. Let me talk with the team and I'll get back to you.

**Reply by jakebailey** [reply]
> Ah, this is super interesting. I didn't realize people worked on one package at a time in the editor rather than working with the entire repo. I'll admit my knowledge of managing monorepos is fairly elementary as I've never had to use one. So it sounds like you're saying that each package is effectively maintained separately without any overriding management?

More or less; think of DT as "Wikipedia for types". It's a monorepo with a load of automation to make it so that most changes do not require the maintainers to review things, only specific package "owners" who volunteer to manage some `@types` packages. Some contributors may look at the whole repo to make one change, or look at the whole repo to touch related packages, but many just open the one dir they work on. We even provide instructions such that they can clone _just_ the directory they're working on and not the other 9000 packages.

This kind of sparseness isn't always true of monorepos in general, though; I think most are more interconnected than that.

> The problem with this approach is that we then need to search for all three files in every directory that we search (three `fs.stat()` calls). That's why for eslintrc we just took the first one that was found.

The way I'd do it is to use `readdir` instead; that will give you a list of files in a dir such that you can error when there are multiple, then do a stat call once if there is just one. That or use `fs.readdirSync(dir, { withFileTypes: true })`, which will give `fs.Dirents` back directly, which I believe may be no slower. Would need a benchmark, of course.

For example, this is the code I've written for `hereby`: https://github.com/jakebailey/hereby/blob/c71a57bcba83f1ad780fdb32e82914ec275bba94/src/cli/loadHerebyfile.ts#L16

I have not tested to see if `withFileTypes` is faster, though.

> That said, I think some good points we hadn't considered were mentioned in this thread. Let me talk with the team and I'll get back to you.

Thank you, I appreciate it.

**Reply by jakebailey** [reply]
Now, I should say that the `readdir` method may actually be _worse_ in that it'll return back 9000 things for our main `types` folder. But, it hasn't caused a problem before (but maybe eslint currnetly stats everything indidually).

**Reply by jakebailey** [reply]
Hm, `withFileTypes` appears to `stat`, so scratch that. Unfortunate; I guess not all OSs have the ability to provide more info on readdir directly. It's optimal in other languages :(

In any case, checking names first is enough to reduce it to one stat call (but again perf testing would be needed).

I am totally happy to prototype some of this if you'd like.

**Reply by jakebailey** [reply]
Sorry, I'm waffling. `withFileTypes` does use the info provided by `readdir` when available (`getDirent` has parametes for path/name/type); my reading comprehension was apparently poor; so that's back on the table.

**Comment by nzakas** [comment]
@karlhorky we don't maintain the VS Code extension, so I can't comment on how it works.

I think your numbers are somewhat opaque, as I don't believe the number of ESM-only packages is an indicator of anything in particular, except perhaps people who don't want to have a build step to generate a CommonJS entrypoint.

**Comment by nzakas** [comment, marked as answer]
Popping back out to the top of the stack here. After discussing with the team, we've agreed that the DT use case is a compelling argument to allow the alternate filenames of `eslint.config.mjs` and `eslint.config.cjs` by default. We are working through the details of what that looks like so stay tuned to the [issue](https://github.com/eslint/eslint/issues/17863).

**Comment by jakebailey** [comment]
Coming back to this, I've now begun trying to convert the tooling to flat config, only to discover that via #16960, you can't reasonably have more that one config file within a repo; only one of them is used depending on cwd. This means that even though `eslint.config.mjs` is now a consistent filename I could stick in all monorepo packages, if I do that, my config is inconsistent depending on how I invoke eslint. The result is that if I want to get things working, I have to give up on having a config for each package that needs it, and instead invent a bespoke configuration method and have a root `eslint.config.js` file read all 1000 config files from the child dirs in order to build a single config. Then since we invoke eslint once per package.... that's quite literally `9000*1000` extra config parses.
